Feature/commonjs

EddyLane opened the pull request:
This refactors our app into CommonJS modules, addressing #31 

Global modules currently needed to run npm commands:

**npm install -g browserify**
**npm install -g uglify**
**npm install -g watch**

You'll need to install the npm requirements with **npm install**.

To run the watch task (watches and re-does browserify):

**npm run-script watch**

To run the build task (browserify and uglify):

**npm run-script build:js**

These tasks could do with being improved as I just did them very much last minute, but should be an okay foundation to start us off.

the **main.js** file now has a global application router, which dictates which code that particular page is asking for.

The only issue I had was with the list.js dep @ryanhyslop added for constituency search, as this has no NPM install unfortunately. May need to fork this one.

I'll try and write some more information tomorrow but I thought we need to get this in as soon as possible to avoid any conflicts that may arise. 

-Eddy

----

- [x] Check out branch, npm install
- [x] Test `npm run build:js`
- [x] Test `npm run watch:js`
- [x] Test site in Chrome
- [ ] Test site in Safari
- [ ] Test site in Firefox
- [ ] Add 'use strict'; (new branch after merging into master)

\- Simon

simonplend commented:
This is great to see - I think it will really make the JS a pleasure to work with!

I'm going to pause work on the DocPad branch (https://github.com/voteforpolicies/data-website/pull/33) as there's not much left to do and I'll face even more merge conflicts if I just plough ahead. Once this Browserify branch is in master (this evening?), I will update my DocPad branch and finish the templating.

We should be in great shape by then :)

simonplend commented on package.json:
```diff
@@ -4,7 +4,9 @@
-    "test": "./node_modules/karma/bin/karma start"
+    "test": "./node_modules/karma/bin/karma start",
+    "build:js": "browserify js/main.js | uglifyjs > dist/js/build.js",
+    "watch:js": "watch 'browserify js/main.js -o dist/js/build.js' js"
```
I'd suggest passing the `-d` switch to `browserify` in this task so that we have source maps in development.

`watchify` might be a bit more flexible and faster (default 100ms polling interval, whereas watch seems to poll in the order of seconds): https://www.npmjs.com/package/watchify

simonplend commented on package.json:
```diff
@@ -22,10 +24,21 @@
   },
   "homepage": "https://github.com/voteforpolicies/data-website",
   "devDependencies": {
+    "jshint": "latest",
```
For the future, but would be good to get our own `.jshintrc` file in the repo at some point.

simonplend commented on package.json:
```diff
@@ -22,10 +24,21 @@
+  },
+  "dependencies": {
+    "chartist": "^0.7.3",
+    "d3": "^3.5.5",
```
This nails this issue: https://github.com/voteforpolicies/data-website/issues/3

simonplend commented on package.json:
```diff
@@ -22,10 +24,21 @@
+    "jquery": "^2.1.3",
+    "lodash": "^3.6.0",
+    "nimble": "0.0.2",
+    "topojson": "^1.6.19"
```
This nails this issue: https://github.com/voteforpolicies/data-website/issues/4

simonplend commented on js/constituencies/results.js:
```diff
@@ -1,132 +1,139 @@
+var nimble = require('nimble');
+var Handlebars = require('Handlebars');
+var Chartist = require('chartist');
+var CONFIG = require('../config');
```
I've noticed that in most other places this is still called `VFP_CONFIG_DATA` (likely a vestige of the old structure). Would be good to pick one variable name for the config and use it consistently.

EddyLane commented:
Cool, i've updated a bit to fix the issues outlined here.

- All npm tasks now running from local dependencies which should help as we'll all be on same versions and won't need to install a bunch of globals
- Now using watchify
- Added sourcemaps when watching
- Slight refactor on autocomplete. Still a bit annoying list.js isnt available as an npm module!

simonplend commented on .jshintrc:
```diff
@@ -0,0 +1,25 @@
+  "curly": true,
+  "eqeqeq": true,
+  "immed": true,
+  "indent": 2,
```
Indent is 4

simonplend commented on .jshintrc:
```diff
@@ -0,0 +1,25 @@
+{
+  "node": true,
+  "esnext": true,
```
Are we actually using ES6?

simonplend commented on .jshintrc:
```diff
@@ -0,0 +1,25 @@
+  "node": true,
+  "esnext": true,
+  "bitwise": true,
+  "camelcase": false,
```
This option is deprecated (see: http://jshint.com/docs/options/). On a related note, we could later look at using JSCS too.

simonplend commented on .jshintrc:
```diff
@@ -0,0 +1,25 @@
+  "camelcase": false,
+  "curly": true,
+  "eqeqeq": true,
+  "immed": true,
```
Also deprecated.

simonplend commented on .jshintrc:
```diff
@@ -0,0 +1,25 @@
+  "immed": true,
+  "indent": 2,
+  "latedef": true,
+  "newcap": true,
```
Also deprecated.

simonplend commented on .jshintrc:
```diff
@@ -0,0 +1,25 @@
+  "latedef": true,
+  "newcap": true,
+  "noarg": true,
+  "quotmark": "single",
```
Also deprecated, but happy to keep this here until/if we add JSCS.

simonplend commented on .jshintrc:
```diff
@@ -0,0 +1,25 @@
+  "undef": true,
+  "unused": true,
+  "strict": false,
+  "trailing": true,
```
I don't see this option.

simonplend commented on .jshintrc:
```diff
@@ -0,0 +1,25 @@
+  "strict": false,
+  "trailing": true,
+  "smarttabs": true,
+  "white": true,
```
I can't find the options `trailing`, `smarttabs` or `white` in the docs (http://jshint.com/docs/options/).

EddyLane commented:
This is just a stock jshintrc I had lying around that came from a Yeoman generator, admittedly it was an ES6 project. I think for now i'll remove it and we can start a discussion if you want over starting a new one, just wanted to get something in there.
